### PR TITLE
Plugin commands clear caches using normalized cache clearing workflow

### PR DESCRIPTION
**Pull request by JoshuaBehrens**
### 1. Why is this change necessary?
There is [a hint](https://github.com/shopware/shopware/blob/f32eba00c30de2f4098a6f5083879b63ad6e597c/engine/Shopware/Commands/PluginActivateCommand.php#L90) that I should clear the cache after installing a plugin. Why not clear it directly?

### 2. What does this change do, exactly?
The scheduled caches from the plugin installation change are passed to the backend cache controller. This is replicated in the commands to change a plugin installation state. Both workflows are using the same methods, which are now extracted into the cache service.

### 3. Describe each step to reproduce the issue or behaviour.
1. Activate a plugin that adds a feature to the storefront via command line
2. Reload page
3. See no changes
4. Facepalm because one forgot the cache clear which is done by the backend automatically
5. `sw:cache:clear`

### 4. Which documentation changes (if any) need to be made because of this PR?
`--clear-cache` is a new option for some plugin commands to enable cache clearing as this is now part of the routines. This might change scripts in the composer-project. There is a new cache tag: search.

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [x] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

### CODE REVIEW

**Comment by JoshuaBehrens** [comment]
I am not sure whether the travis errors are part of my PR as they seem to be unrelated.

**Comment by t2oh4e** [comment]
As this differs from the current implementation and from the backend workflow (you have to explicitly clear the caches. It does not happen automatically.) I would prefer to turn it around: Add a flag to explicitly clear the necessary caches instead of adding a flag to disable an automatic cache clear.
Furthermore the linked hint does not tell you that you should clear the caches but that you may want to do it ;)
Any other opinions on that?

**Comment by JoshuaBehrens** [comment]
@t2oh4e Good point that cache clearing is optional in the backend. This change seems reasonable to me. Going to change that.

File: engine/Shopware/Components/CacheManager.php [hunk 1]
```diff
@@ -447,6 +461,61 @@ public function encodeSize($bytes)
...
+     * @param string[] $tags
+     *
+     * @return int
+     */
+    public function clearByTags(array $tags)
+    {
+        $count = 0;
+
+        foreach (array_unique($tags) as $tag) {
+            $count += (int) $this->clearByTag($tag);
```
Comment (hlohaus) [inline engine/Shopware/Components/CacheManager.php]: Die Rückgabe von clearByTag ist void oder false.

**Comment by mitelg** [comment]
@JoshuaBehrens nice work! :+1: 
but please adjust your initial description of the PR, as you did some more changes, than just the commands :wink:

File: engine/Shopware/Components/CacheManager.php (hunk 1 above)
Comment (JoshuaBehrens) [reply engine/Shopware/Components/CacheManager.php]: I don't see a mixed return type. All I see that I always return false which is wrong. The last return should've been true.

**Comment by JoshuaBehrens** [comment]
Is everyone fine with the recent changes?

**Comment by jkrzefski** [comment]
I added a summary of which caches have been cleared successfully and which have failed to be cleared. Also I included some minor bug-fixes and made a PR to this PR, so my changes are isolated and can be reviewed separately. Take a look at this one: https://github.com/niemand-online/shopware/pull/1

File: engine/Shopware/Commands/PluginActivateCommand.php [hunk 2]
```diff
@@ -85,8 +93,16 @@ protected function execute(InputInterface $input, OutputInterface $output)
             return 1;
         }
 
-        $pluginManager->activatePlugin($plugin);
+        $activationContext = $pluginManager->activatePlugin($plugin);
+        $output->writeln(sprintf('Plugin %s has been activated.', $pluginName));
 
-        $output->writeln(sprintf('Plugin %s has been activated. Consider sw:cache:clear to enable possible behaviors that come with the plugin.', $pluginName));
+        if (!empty($input->getOption('clear-cache'))) {
```
Comment (soebbing) [inline engine/Shopware/Commands/PluginActivateCommand.php]: You could turn this into an early return

Comment (jkrzefski) [reply engine/Shopware/Commands/PluginActivateCommand.php]: @soebbing Would you keep the message about the scheduled cache clearing in case it is not dealt with automatically like so?

```php
if (empty($input->getOption('clear-cache'))) {
    $output->writeln(sprintf('Plugin %s has been activated. Consider sw:cache:clear to enable possible behaviors that come with the plugin.', $pluginName));
    return 0;
}
```

Comment (soebbing) [reply engine/Shopware/Commands/PluginActivateCommand.php]: @jkrzefski Personally, I think that is a nice addition

File: engine/Shopware/Components/CacheManager.php (hunk 1 above)
Comment (hlohaus) [reply engine/Shopware/Components/CacheManager.php]: Ich meine nur, macht das dann mit den Zahlen addieren sinn?

Comment (jkrzefski) [reply engine/Shopware/Components/CacheManager.php]: The initial idea of this was to return the amount of successfully cleared caches. Personally I do not think of this feature as very useful. That's why I changed the return of this method from an integer (how many caches were cleared successfully) to an array of strings (which caches were cleared successfully) in my pr to this pr (see https://github.com/niemand-online/shopware/pull/1 ).

TL;DR: I do agree with you that this implementation does not make much sense.

**Comment by JoshuaBehrens** [comment]
I added more changes heavily inspired by @jkrzefski PR (which does not render obsolete) and @soebbing recent comment.

**Comment by jkrzefski** [comment]
Before this PR gets merged, please take a look at https://github.com/niemand-online/shopware/pull/1 which will add a summary output to console commands. Also that PR will add support for legacy plugins so I would strongly suggest to include it in this PR.

**Comment by JoshuaBehrens** [comment]
Done @shopwareBot

File: engine/Shopware/Controllers/Backend/Cache.php [hunk 3]
```diff
@@ -131,29 +131,26 @@ public function clearCacheAction()
...
-        }
-        if ($cache['proxy'] === 'on') {
-            $this->cacheManager->clearProxyCache();
-            $this->cacheManager->clearOpCache();
+        $cacheTags = [];
+
+        foreach ($cache as $cacheTag => $value) {
+            if ($value === 'on') {
+                if ($cacheTag === 'frontend') {
+                    $cacheTags[] = CacheManager::CACHE_TAG_CONFIG;
```
Comment (hlohaus) [inline engine/Shopware/Controllers/Backend/Cache.php:139]: I saw it already above? Duplicate Code

Comment (JoshuaBehrens) [reply engine/Shopware/Controllers/Backend/Cache.php:139]: Pardon, to what is it a duplicate to?

File: engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php [hunk 4]
```diff
@@ -341,7 +342,21 @@ private function applyLegacyResultToContext($result, InstallContext $context)
         }
 
         if (array_key_exists('invalidateCache', $result)) {
-            $context->scheduleClearCache($result['invalidateCache']);
+            foreach ($result['invalidateCache'] as $cacheTag) {
```
Comment (hlohaus) [inline engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php:345]: here. maybe you can do this in clearByTag

Comment (jkrzefski) [reply engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php:345]: Similar indeed, but not the same. To combine these in a common method seems possible, but I am unsure if more changes from us would have any effect now, since the pr is already closed.

Comment (hlohaus) [reply engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php:345]: mhh. Shopware seems to have forgotten it:
https://github.com/shopware/shopware/commit/99369a56b0bc5cc472c22c174ead25e02d2504de
Is that bad?

Comment (jkrzefski) [reply engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php:345]: As long as it works, I don't think it is bad. It is just one thing that *could* be better. Maybe you can add another PR on top of [this commit](https://github.com/shopware/shopware/commit/99369a56b0bc5cc472c22c174ead25e02d2504de)?

Comment (JoshuaBehrens) [reply engine/Shopware/Bundle/PluginInstallerBundle/Service/InstallerService.php:345]: These are two different sources of tags. A backend cache manager way of handling tags and a plugin manager way of handling tags. Although they are similar they are not equal yet. I would be happy if they are the same.
